release blog post for Rust 1.46.0 (#675)
rust-lang/blog.rust-lang.org: posts/2020-08-27-Rust-1.46.0.md

---
layout: post
title: "Announcing Rust 1.46.0"
author: The Rust Release Team
release: true
---

The Rust team is happy to announce a new version of Rust, 1.46.0. Rust is a
programming language that is empowering everyone to build reliable and
efficient software.

If you have a previous version of Rust installed via rustup, getting Rust
1.46.0 is as easy as:

```console
rustup update stable
```

If you don't have it already, you can [get `rustup`][install] from the
appropriate page on our website, and check out the [detailed release notes for
1.46.0][notes] on GitHub.

> [install]: https://www.rust-lang.org/install.html

Reviewer: A minor thing, but it looks like this page is now /tools/install:

```suggestion
[install]: https://www.rust-lang.org/tools/install
```

[notes]: https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1460-2020-08-27

## What's in 1.46.0 stable

> This release is on the smaller side, with a number of improvements to `const

Reviewer 2: const fn improvements are really big to me! It's not a small release.

Author: Yes, as mentioned above, I'm changing this. I did not realize how much power this gave folks; and it's a pretty small *number* of things, so I mis-judged!

fn`, two new standard library APIs, and one feature useful for library
authors. See the [detailed release notes][notes] to learn about other changes
not covered by this post.

### `#[track_caller]`

Back in March, the release of Rust 1.42 brought about [better error messages when unwrap would panic][better-errors]. At the time, we mentioned that the way
this was implemented was not yet stable. Rust 1.46 stabilizes this feature.

[better-errors]: https://blog.rust-lang.org/2020/03/12/Rust-1.42.html#useful-line-numbers-in-option-and-result-panic-messages

This attribute is called `#[track_caller]`, which was originally proposed
in [RFC 2091][rfc-2091] way back in July of 2017! If you're writing a function
like unwrap, where you want errors to report to the caller, you can put this
annotation on your functions, and the default panic formatter will use it to

> print its error message. For example, here is `unwrap` previously:

Author: I've worked something else up, thanks for giving it a shot! Will push in a few minutes.

```rust
pub fn unwrap(self) -> T {
    match self {
        Some(val) => val,
        None => panic!("called `Option::unwrap()` on a `None` value"),
    }
}
```

It now looks like this:

```rust
#[track_caller]
pub fn unwrap(self) -> T {
    match self {
        Some(val) => val,
        None => panic!("called `Option::unwrap()` on a `None` value"),
    }
}
```

> That's it!

Reviewer 3: ...though an example that uses `panic::Location` might be helpful.

> That's it!

Author: I wanted to emphasize how simple this is to use, but if you wanted to come up with an example using Location, I'm not inherently opposed. I didn't show how the output was changed because we did that when it was added to unwrap.

If you are implementing a panic hook yourself, you can use the [caller] method
on `std::panic::Location` to get access to this information.

[rfc-2091]: https://github.com/rust-lang/rfcs/pull/2091
[caller]: https://doc.rust-lang.org/stable/std/panic/struct.Location.html#method.caller

### `const fn` improvements

> You can now use [a variety of things] in a `const fn`:

Reviewer 3: ```suggestion
There are [several core language features] you can now use in a `const fn`:
```

> * `if`
> * `match`
> * the `&&` and `||` operators
> * `loop`

Reviewer 4: Would it be useful to call out all the language constructs that were stabilized? That is:

```suggestion
* `if`, `if let`, and `match`
* `while`, `while let`, and `loop`
* the `&&` and `||` operators
```

Author: Yes! I was going by the release notes, so I didn't know it was all of these things. Thank you!

You can also [cast to a slice][cast-to-slice]:

```rust
const fn foo() {
  let x = [1, 2, 3, 4, 5];

  // cast the array to a slice
  let y: &[_] = &x;
}
```

> [a variety of things]: https://github.com/rust-lang/rust/pull/72437/

Reviewer 3: ```suggestion
[several core language features]: https://github.com/rust-lang/rust/pull/72437/
```
